### ISSUE TITLE
using stdout or stderr causes segmentation fault on Linux i386 target

### ISSUE THREAD

**Issue by pts** [open-watcom/open-watcom-v2#1053]
Source file *stdiotest.c*:

```c
#include <stdio.h>
int main(int argc, char **argv) {
  (void)argc; (void)argv;
  fprintf(stdout, "OK1.\n");
  fflush(stdout);
  if (ferror(stdout)) return 1;
  fprintf(stderr, "OK2.\n");
  fflush(stderr);
  if (ferror(stderr)) return 2;
  return 0;
}
```

Reproduce the error:

```
$ gcc -s -O2 -W -Wall -o stdiotest stdiotest.c && ./stdiotest; echo $?
OK1.
OK2.
0
$ owcc -blinux -s -O2 -W -Wall -o stdiotest.ow stdiotest.c && ./stdiotest.ow; echo $?
Segmentation fault (core dumped)
139
```

Workaround: if I use `fdopen(1, "w")` instead of `stdout` (and likewise `fdopen(2, "w")` instead of `stderr`), then it works with both `gcc` and `owcc`.

**Comment by jmalak** [comment]
Thanks for your report.
I am sorry but I cannot reproduce this problem either with 64-bit or 32-bit version on Linux (I used latest build snapshot).
Please specify your platform and how you have have configured Open Watcom (environment variables setup).

**Comment by pts** [comment]
Thank you for looking into this!

I can reproduce this segmentation fault consistently on my Linux amd64 system. I've also deleted all environment varialbes (with `env -i`) to make it easier to reproduce. I've also attached the *stdiotest.ow* executable program. Could you please also attach your *stdiotest.ow*, so that I can try it on my system?

[stdiotest.ow.zip](https://github.com/open-watcom/open-watcom-v2/files/11003059/stdiotest.ow.zip)

```
$ uname -a
Linux myhost 5.4.0-124-generic #140~18.04.1-Ubuntu SMP Fri Aug 5 11:43:34 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ export WATCOM="$HOME/Downloads/open-watcom-2"; env -i WATCOM="$WATCOM" INCLUDE="$WATCOM/h" PATH="$WATCOM/binl:/bin:/usr/bin" owcc -v -fd -blinux -s -O2 -W -Wall -o stdiotest.ow stdiotest.c && env -i ./stdiotest.ow; echo $?
Open Watcom C/C++ x86 32-bit Compiler Driver Program
Version 2.0 beta Mar  4 2023 01:01:17 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
        wcc386 -bt=linux -onatx -W -w4 -3r -fo=.o -fr stdiotest.c 
Open Watcom C x86 32-bit Optimizing Compiler
Version 2.0 beta Mar  4 2023 00:54:39 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
stdiotest.c: 12 lines, included 814, 0 warnings, 0 errors
Code size: 89
        wlink @__owcc00.lnk 
Open Watcom Linker Version 2.0 beta Mar  4 2023 00:51:12 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
creating a Linux x86 executable
Segmentation fault (core dumped)
139
$ cat __owcc__.lnk 
system linux
file stdiotest.o
name stdiotest.ow
option noextension
$ env -i strace ./stdiotest.ow 
execve("./stdiotest.ow", ["./stdiotest.ow"], 0x7ffe776048a0 /* 0 vars */) = 0
strace: [ Process PID=12055 runs in 32 bit mode. ]
brk(NULL)                               = 0x91ed000
brk(0x91f5000)                          = 0x91f5000
gettid()                                = 12055
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0xd068000c} ---
+++ killed by SIGSEGV (core dumped) +++
Segmentation fault (core dumped)
```

> Anyway what is purpose of `-W` option?

It looks like it's a no-op for *owcc* and *wcc*. I use it with *gcc* to enable additional warnings. Feel free to drop it.

**Comment by jmalak** [comment]
Thanks for all info.
What is now clear from your information is that problem appears with linker.
I will try to run resulting executable.
please could you add option **`-fm`** to owcc command line to create linker map file for compare.

**Comment by jmalak** [comment]
I will try to compile it ones more and check run-time if on my box it is working or not and note you about result to be able somehow identify root cause. 
I am using Ubuntu 22.04 with GCC 12.1.0.

**Comment by pts** [comment]
Thank you for the uploads!

**TL;DR** I've found the culprint: it's `#pragma pack` in OpenWatcom `h/stdio.h`.

For me it still segfaults when I compile it on Linux with [releases/tag/2023-04-01-Build](https://github.com/open-watcom/open-watcom-v2/releases/tag/2023-04-01-Build) (both [open-watcom-2_0-c-linux-x64](https://github.com/open-watcom/open-watcom-v2/releases/download/2023-04-01-Build/open-watcom-2_0-c-linux-x64) and [open-watcom-2_0-c-linux-x86](https://github.com/open-watcom/open-watcom-v2/releases/download/2023-04-01-Build/open-watcom-2_0-c-linux-x86)). Here is how I compile and run it:

The following attachment contains the stdiotest.c, the generated Linux executable program stdiotest.ow and the generated map file stdiotest.map: [stdiotest.zip](https://github.com/open-watcom/open-watcom-v2/files/11259532/stdiotest.zip)

```
$ uname -a
Linux myhost 5.4.0-124-generic #140~18.04.1-Ubuntu SMP Fri Aug 5 11:43:34 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ export WATCOM="$HOME/Downloads/open-watcom-2_0-c-linux-x86-2023-04-01"; env -i WATCOM="$WATCOM" INCLUDE="$WATCOM/h" PATH="$WATCOM/binl:/bin:/usr/bin" owcc -fm -v -fd -blinux -s -O2 -W -Wall -o stdiotest.ow stdiotest.c && ls -ld ./stdiotest.ow && env -i ./stdiotest.ow; echo $?
Open Watcom C/C++ x86 32-bit Compiler Driver Program
Version 2.0 beta Apr  1 2023 01:12:04 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
        wcc386 -bt=linux -onatx -W -w4 -3r -fo=.o -fr stdiotest.c 
Open Watcom C x86 32-bit Optimizing Compiler
Version 2.0 beta Apr  1 2023 01:05:27 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
stdiotest.c: 12 lines, included 814, 0 warnings, 0 errors
Code size: 89
        wlink @__owcc00.lnk 
Open Watcom Linker Version 2.0 beta Apr  1 2023 01:02:06 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
creating map file
creating a Linux x86 executable
-rwxrwxr-x 1 pts pts 23150 Apr 18 09:19 ./stdiotest.ow
Segmentation fault (core dumped)
139
```

The stdiotest.ow program you have posted in https://github.com/open-watcom/open-watcom-v2/issues/1053#issuecomment-1480074720 works correctly (*OK1* and *OK2*) on my Linux system. I've renamed the working one to stdiotest.ow.good, and compared to the one which produces the segmentation fault. The size is the same, only 6 bytes are different:

```
$ cmp -l stdiotest.ow.good stdiotest.ow
  279  54  52
  292  54  52
  303  70  66
  322 110 104
  335 110 104
  346 124 120
```

The corresponding hex differences are:

```
at 0x116 change 0x2c to 0x2a
at 0x123 change 0x2c to 0x2a
at 0x12e change 0x38 to 0x36
at 0x141 change 0x48 to 0x44
at 0x14e change 0x48 to 0x44
at 0x159 change 0x54 to 0x50
```

By looking it up in the .map file, it looks like all these differences are in the *main* function:

```
at 0x116==main_+0x06 change 0x2c to 0x2a
at 0x123==main_+0x13 change 0x2c to 0x2a
at 0x12e==main_+0x1e change 0x38 to 0x36
at 0x141==main_+0x31 change 0x48 to 0x44
at 0x14e==main_+0x3e change 0x48 to 0x44
at 0x159==main_+0x49 change 0x54 to 0x50
```

Disassembling the bad *main* function:

```
$ export WATCOM="$HOME/Downloads/open-watcom-2_0-c-linux-x86-2023-04-01"; env -i WATCOM="$WATCOM" INCLUDE="$WATCOM/h" PATH="$WATCOM/binl:/bin:/usr/bin" owcc -c -v -fd -blinux -O2 -W -Wall -o stdiotest.obj stdiotest.c && wdis stdiotest.obj
Open Watcom C/C++ x86 32-bit Compiler Driver Program
Version 2.0 beta Apr  1 2023 01:12:04 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
        wcc386 -bt=linux -onatx -W -w4 -3r -fo=stdiotest.obj -fr stdiotest.c 
Open Watcom C x86 32-bit Optimizing Compiler
Version 2.0 beta Apr  1 2023 01:05:27 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
stdiotest.c: 12 lines, included 814, 0 warnings, 0 errors
Code size: 89
Module: /home/pts/Downloads/openwatcom_try/stdiotest/stdiotest.c
GROUP: 'DGROUP' CONST,CONST2,_DATA

Segment: _TEXT PARA USE32 00000059 bytes
0000                            main_:
0000  68 00 00 00 00                    push            offset L$3
0005  68 1A 00 00 00                    push            offset ___iob+0x1a
000A  E8 00 00 00 00                    call            fprintf_
000F  83 C4 08                          add             esp,0x00000008
0012  B8 1A 00 00 00                    mov             eax,offset ___iob+0x1a
0017  E8 00 00 00 00                    call            fflush_
001C  F6 05 26 00 00 00 20              test            byte ptr ___iob+0x26,0x20
0023  74 06                             je              L$1
0025  B8 01 00 00 00                    mov             eax,0x00000001
002A  C3                                ret
002B                            L$1:
002B  68 06 00 00 00                    push            offset L$4
0030  68 34 00 00 00                    push            offset ___iob+0x34
0035  E8 00 00 00 00                    call            fprintf_
003A  83 C4 08                          add             esp,0x00000008
003D  B8 34 00 00 00                    mov             eax,offset ___iob+0x34
0042  E8 00 00 00 00                    call            fflush_
0047  F6 05 40 00 00 00 20              test            byte ptr ___iob+0x40,0x20
004E  74 06                             je              L$2
0050  B8 02 00 00 00                    mov             eax,0x00000002
0055  C3                                ret
0056                            L$2:
0056  31 C0                             xor             eax,eax
0058  C3                                ret

Routine Size: 89 bytes,    Routine Base: _TEXT + 0000

No disassembly errors

Segment: CONST DWORD USE32 0000000C bytes
0000                            L$3:
0000  4F 4B 31 2E 0A 00                               OK1...
0006                            L$4:
0006  4F 4B 32 2E 0A 00                               OK2...

Segment: CONST2 DWORD USE32 00000000 bytes

Segment: _DATA DWORD USE32 00000000 bytes
```

It looks like *main* gets the offsets within *__iob* wrong:

```
0005  68 1A 00 00 00                    push            offset ___iob+0x1a  ; stdout. Correct would be ___iob+0x1c.
0012  B8 1A 00 00 00                    mov             eax,offset ___iob+0x1a  ; stdout. Correct would be ___iob+0x1c.
001C  F6 05 26 00 00 00 20              test            byte ptr ___iob+0x26,0x20  ; stdout->_flag. Correct would be ___iob+0x28.
0030  68 34 00 00 00                    push            offset ___iob+0x34  ; stderr. Correct would be ___iob+0x38.
003D  B8 34 00 00 00                    mov             eax,offset ___iob+0x34  ; stderr. Correct would be ___iob+0x38.
0047  F6 05 40 00 00 00 20              test            byte ptr ___iob+0x40,0x20  ; stderr->_flag. Correct would be ___iob+0x44.
```

To diagnose this further, I've run stdiotest.c through the C preprocessor (`owcc -E`):

```
$ export WATCOM="$HOME/Downloads/open-watcom-2_0-c-linux-x86-2023-04-01"; env -i WATCOM="$WATCOM" INCLUDE="$WATCOM/h" PATH="$WATCOM/binl:/bin:/usr/bin" owcc -E -v -fd -blinux -O2 -W -Wall -o stdiotest_pre.c stdiotest.c
Open Watcom C/C++ x86 32-bit Compiler Driver Program
Version 2.0 beta Apr  1 2023 01:12:04 (32-bit)
Copyright (c) 2002-2023 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
        wcc386 -bt=linux -onatx -W -w4 -pl -3r -fo=stdiotest.pre.c -fr stdiotest.c 
```

After manually sanitizing the output, I've got this C source code, which still causes Segmentation fault:

```c
/* Causes Segmentation fault. Fix it by removing the 2 `#pragma pack' lines. */
#line 132 "open-watcom-2_0-c-linux-x86-2023-04-01/h/stdio.h"
#pragma pack(__push, 1)
struct __iobuf {
  unsigned char        *_ptr;
  int                   _cnt;
  struct __stream_link *_link;
  unsigned              _flag;
  int                   _handle;
  unsigned              _bufsize;
  unsigned short        _ungotten;
};
typedef struct __iobuf FILE;
typedef FILE __w_FILE;
__declspec(__watcall) extern __w_FILE __near __iob[];
__declspec(__watcall) extern int fprintf(FILE *__fp, const char *__format, ...);
__declspec(__watcall) extern int fflush(FILE *__fp);
#define stdout ((__w_FILE *)&__iob[1])
#define stderr ((__w_FILE *)&__iob[2])
#define ferror(fp) ((fp)->_flag & 0x0020)
#line 556 "open-watcom-2_0-c-linux-x86-2023-04-01/h/stdio.h"
#pragma pack(__pop)
#line 1 "stdiotestb.c"

int main(int argc, char **argv) {
  (void)argc; (void)argv;
  fprintf(stdout, "OK1.\n");
  fflush(stdout);
  if (ferror(stdout)) return 1;
  fprintf(stderr, "OK2.\n");
  fflush(stderr);
  if (ferror(stderr)) return 2;
  return 0;
}
```

When I removed the `#pragma pack` lines, it started working again (*OK1* and *OK2*). Thus with `#pragma pack`, *owcc* produces the buggy executable program (same as *stdiotest.ow.bad*), without the `#pragma pac`, *owcc* produces the correct executable program (same as *stdiotest.ow* attached). **I've found the culprit!**

I'll post some more info here, extending my comment.
